ddapm_test_agent/agent.py

```diff
@@ -187,18 +190,84 @@ async def traces(self) -> TraceMap:
+            if token in self._trace_failures:
+                del self._trace_failures[token]
+            if token in self._trace_check_results_by_check:
+                del self._trace_check_results_by_check[token]
+        log.info(f"Clearing {len(trace_failures)} Trace Check Failures for Token {token}, clear_all={clear_all}")
+        log.info(trace_failures)
+        return web.HTTPOk()
+
     def get_trace_check_failures(self, request: Request) -> web.Response:
-        """Return the Trace Check failures that occurred, if pooling is enabled as a request."""
-        trace_failures = self._trace_failures
-
-        if len(trace_failures) > 0:
-            failure_message = f"APM Test Agent Validation failed with {len(trace_failures)} Trace Check failures.\n"
-            for trace_check_message in trace_failures:
-                failure_message += trace_check_message
-            return web.HTTPBadRequest(text=failure_message)
+        """Return the Trace Check failures that occurred, if pooling is enabled,
+        returned as either a Text (by default) or JSON response.
+        """
+        token = request["session_token"]
+        return_all = "return_all" in request.query and request.query["return_all"].lower() == "true"
+
+        if return_all:
+            # check for whether to return all results
+            trace_check_failures = []
+            for token, f in self._trace_failures.items():
+                trace_check_failures.extend(f)
+            n_failures = len(trace_check_failures)
+            log.info(f"{n_failures} Trace Failures Occurred in Total")
+        else:
+            # or return results by token
+            trace_check_failures = self._trace_failures.get(token, [])
+            n_failures = len(trace_check_failures)
+            log.info(f"{n_failures} Trace Failures Occurred for Token {token}")
+        if n_failures > 0:
+            if "use_json" in request.query and request.query["use_json"].lower() == "true":
+                # check what response type to use
+                results: Dict[str, List[str]] = {}
+                for check_trace, failure_message in trace_check_failures:
+                    results = check_trace.get_failures_by_check(results)
+                json_summary = json.dumps(results)
+                return web.HTTPBadRequest(body=json_summary, content_type="application/json")
+            else:
+                # or use default response of text
+                msg = f"APM Test Agent Validation failed with {n_failures} Trace Check failures.\n"
+                for check_trace, failure_message in trace_check_failures:
+                    msg += failure_message
+                return web.HTTPBadRequest(text=msg)
         else:
             return web.HTTPOk()
 
+    def get_trace_check_summary(self, request: Request) -> web.Response:
+        token = request["session_token"]
+        summary: Dict[str, Dict[str, int]] = {}
```

Comment:
nit: you can use cllections.defaultdict() to clear up some of the if/else statements below

Reply:
thanks! much cleaner now